staking-pool-shares/src/lib.rs

```diff
@@ -0,0 +1,759 @@
+    /// The fraction of the reward that goes to the owner of the staking pool for running the
+    /// validator node.
+    pub reward_fee_fraction: RewardFeeFraction,
+    /// Persistent map of the account ID to the account.
+    pub accounts: Map<AccountId, Account>,
+}
+
+impl Default for StakingContract {
+    fn default() -> Self {
+        env::panic(b"Staking contract should be initialized before usage")
+    }
+}
+
+#[derive(BorshDeserialize, BorshSerialize, Serialize, Deserialize, Clone)]
+pub struct RewardFeeFraction {
+    pub numerator: u32,
+    pub denominator: u32,
+}
+
+impl RewardFeeFraction {
+    pub fn assert_valid(&self) {
+        assert_ne!(self.denominator, 0, "Denominator must be a positive number");
+        assert!(
+            self.numerator <= self.denominator,
+            "The reward fee must be less or equal to 1"
+        );
+    }
+
+    pub fn multiply(&self, value: Balance) -> Balance {
+        (U256::from(self.numerator) * U256::from(value) / U256::from(self.denominator)).as_u128()
+    }
+}
+
+#[near_bindgen]
+impl StakingContract {
+    /// Initializes the contract with the given owner_id, initial staking public key (with ED25519
+    /// curve) and initial reward fee fraction that owner charges for the validation work.
+    #[init]
+    pub fn new(
+        owner_id: AccountId,
+        stake_public_key: Base58PublicKey,
+        reward_fee_fraction: RewardFeeFraction,
+    ) -> Self {
+        assert!(!env::state_exists(), "Already initialized");
+        reward_fee_fraction.assert_valid();
+        assert!(
+            env::is_valid_account_id(owner_id.as_bytes()),
+            "The owner account ID is invalid"
+        );
+        Self {
+            owner_id,
+            stake_public_key: stake_public_key.into(),
+            last_epoch_height: env::epoch_height(),
+            last_locked_account_balance: 0,
+            last_account_balance: env::account_balance(),
+            total_staked_balance: 0,
+            total_stake_shares: 0,
+            reward_fee_fraction,
+            accounts: Map::new(b"u".to_vec()),
+        }
+    }
+
+    /// Call to distribute rewards after the new epoch. It's automatically called before every
+    /// action.
+    pub fn ping(&mut self) {
+        // Checking if we need there are rewards to distribute.
+        let epoch_height = env::epoch_height();
+        if self.last_epoch_height == epoch_height {
+            return;
+        }
+        self.last_epoch_height = epoch_height;
+
+        // New total amount (both locked and unlocked balances).
+        // NOTE: We need to subtract `attached_deposit` in case `ping` called from `deposit` call.
+        let total_balance =
+            env::account_locked_balance() + env::account_balance() - env::attached_deposit();
+        // Old total amount.
+        let last_total_balance = self.last_account_balance + self.last_locked_account_balance;
+
+        // TODO: Verify if the reward can become negative.
```

Comment:
Currently it cannot. With slashing it can.

Reply:
Then we can assert here

Reply:
yes